# Initial Data Analysis of German Credit Score Dataset

This is a notebook for exploring German Credit Score data for practice case in day4 bootcamp @IYKRA

**Context**
<br>
The original dataset contains 1000 entries with 20 categorial/symbolic attributes prepared by Prof. Hofmann. In this dataset, each entry represents a person who takes a credit by a bank. Each person is classified as good or bad credit risks according to the set of attributes. The link to the original dataset can be found below.
<br>
**Content**
It is almost impossible to understand the original dataset due to its complicated system of categories and symbols. Thus, I wrote a small Python script to convert it into a readable CSV file. Several columns are simply ignored, because in my opinion either they are not important or their descriptions are obscure. The selected attributes are:

- Age (numeric)
- Sex (text: male, female)
- Job (numeric: 0 - unskilled and non-resident, 1 - unskilled and resident, 2 - skilled, 3 - highly skilled)
- Housing (text: own, rent, or free)
- Saving accounts (text - little, moderate, quite rich, rich)
- Checking account (numeric, in DM - Deutsch Mark)
- Credit amount (numeric, in DM)
- Duration (numeric, in month)
- Purpose(text: car, furniture/equipment, radio/TV, domestic appliances, repairs, education, business, vacation/others
- Risk (Value target - Good or Bad Risk)

The original dataset can be found [here at UCI website](https://archive.ics.uci.edu/ml/datasets/Statlog+%28German+Credit+Data%29). 

In [1]:
# Importing the libraries

import pandas as pd
import numpy as np
import os as os

In [7]:
# Read the dataset and show some samples

df_gc = pd.read_csv('./german_credit_data_with_risk.csv', index_col='idx')
display(df_gc.head())
display(df_gc.tail())

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
idx                                                                             
0     67    male    2     own             NaN           little           1169   
1     22  female    2     own          little         moderate           5951   
2     49    male    1     own          little              NaN           2096   
3     45    male    2    free          little           little           7882   
4     53    male    2    free          little           little           4870   

     Duration              Purpose  Risk  
idx                                       
0           6             radio/TV  good  
1          48             radio/TV   bad  
2          12            education  good  
3          42  furniture/equipment  good  
4          24                  car   bad

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
idx                                                                             
995   31  female    1     own          little              NaN           1736   
996   40    male    3     own          little           little           3857   
997   38    male    2     own          little              NaN            804   
998   23    male    2    free          little           little           1845   
999   27    male    2     own        moderate         moderate           4576   

     Duration              Purpose  Risk  
idx                                       
995        12  furniture/equipment  good  
996        30                  car  good  
997        12             radio/TV  good  
998        45             radio/TV   bad  
999        45                  car  good

In [28]:
# List all columns and their type

df_gc.dtypes

Age                  int64
Sex                 object
Job                  int64
Housing             object
Saving accounts     object
Checking account    object
Credit amount        int64
Duration             int64
Purpose             object
Risk                object
dtype: object

## Part A
<br>
On your Jupyter notebook, answer the following questions!

**1. How many rows and columns reflected in the dataset?**

In [9]:
print('The dataset have {} rows, and {} columns'.format(df_gc.shape[0], df_gc.shape[1]))

The dataset have 1000 rows, and 10 columns


**2. What is the minimum, maximum, and mean of the customer's age?**

In [26]:
print('The minimum age of a customer is {:.0f}'.format(df_gc['Age'].describe()['min']))
print('The maximum age of a customer is {:.0f}'.format(df_gc['Age'].describe()['max']))
print('The mean age of a customer is {:.2f}'.format(df_gc['Age'].describe()['mean']))

The minimum age of a customer is 19
The maximum age of a customer is 75
The mean age of a customer is 35.55


**3. How many are male and how many are female? Why?**

In [35]:
print('Count number of male customer is {}'.format(df_gc['Sex'].value_counts()['male']))
print('Count number of female customer is {}'.format(df_gc['Sex'].value_counts()['female']))

Count number of male customer is 690
Count number of female customer is 310


In [67]:
# Take a look at risk for each sex

tmp1 = df_gc.groupby(by=['Sex','Risk']).count()[['Age']]
print('Percentage of good female to only female customer is {:.2f}%'.format(tmp1.values[0][0]*100/(tmp1.values[0][0]+tmp1.values[1][0])))
print('Percentage of good female to all customer is {:.2f}%'.format(tmp1.values[0][0]*100/tmp1.values.sum()))
print('Percentage of good male to only male customer is {:.2f}%'.format(tmp1.values[3][0]*100/(tmp1.values[2][0]+tmp1.values[3][0])))
print('Percentage of good male to all customer is {:.2f}%'.format(tmp1.values[3][0]*100/tmp1.values.sum()))

Percentage of good female to only female customer is 35.16%
Percentage of good female to all customer is 10.90%
Percentage of good male to only male customer is 72.32%
Percentage of good male to all customer is 49.90%


The data shown that 49.90% of the good customer is male, and 72.32% of the male customer is good

**4. How many high skilled workers customers reflected in the dataset?**

In [74]:
print('The number of customers that are highly skilled workers is {}, which is {:.2f}% from the total customers'
      .format(df_gc['Job'].value_counts()[3], df_gc['Job'].value_counts()[3]*100/df_gc['Job'].value_counts().sum()))

The number of customers that are highly skilled workers is 148, which is 14.80% from the total customers


**5. How many customers who own the house?**

In [82]:
print('The number of customers who own their house is {}, which is {:.2f}% from the total customers'
      .format(df_gc['Housing'].value_counts()['own'], df_gc['Housing'].value_counts()['own']*100/df_gc['Housing'].value_counts().sum()))

The number of customers who own their house is 713, which is 71.30% from the total customers


**6. How many customers who have 'rich' Saving accounts?**

In [83]:
print('The number of customers that categorized as having rich saving accounts is {}, which is {:.2f}% from the total customers'
      .format(df_gc['Saving accounts'].value_counts()['rich'], df_gc['Saving accounts'].value_counts()['rich']*100/df_gc['Saving accounts'].value_counts().sum()))

The number of customers that categorized as having rich saving accounts is 48, which is 5.88% from the total customers


**7. How many customers who do not have checking accounts or have zero checking account?**

In [114]:
# First fill the nan values in Checking accounts column with 'zero' categories
df_gc['Checking account'] = df_gc['Checking account'].fillna('zero')

In [119]:
print('The number of customers that categorized as having zero checking accounts is {}, which is {:.2f}% from the total customers'
      .format(df_gc['Checking account'].value_counts()['zero'], df_gc['Checking account'].value_counts()['zero']*100/df_gc['Checking account'].value_counts().sum()))

The number of customers that categorized as having zero checking accounts is 394, which is 39.40% from the total customers


**8. Who is the one who have highest credit amount? Show us the profile of the age, sex, job, housing, saving account, and checking account!**

In [121]:
# Show data of one customer that have highest credit amount
df_gc.loc[df_gc['Credit amount'] == max(df_gc['Credit amount']), ['Age', 'Sex', 'Job', 'Housing', 'Saving accounts', 'Checking account']]

Age     Sex  Job Housing Saving accounts Checking account
idx                                                           
915   32  female    3     own          little         moderate

**9. How many customers are classified as good risk and bad risk?**

In [123]:
print('Number of customers that classified as good is {}'.format(df_gc['Risk'].value_counts()['good']))
print('Number of customers that classified as bad is {}'.format(df_gc['Risk'].value_counts()['bad']))

Number of customers that classified as good is 700
Number of customers that classified as bad is 300


**10. Is the average of credit amount for good risk is higher than bad risk?**

In [141]:
display(df_gc.groupby(by='Risk').mean()[['Credit amount']])
good_cred_amt = df_gc.groupby(by='Risk').mean()[['Credit amount']].loc['good'].values[0]
bad_cred_amt = df_gc.groupby(by='Risk').mean()[['Credit amount']].loc['bad'].values[0]

if good_cred_amt > bad_cred_amt: print('Yes, the average of credit amount for good risk is higher than bad risk')
else: print('No, the average of credit amount for good risk is lower than bad risk')

Credit amount
Risk               
bad     3938.126667
good    2985.457143

No, the average of credit amount for good risk is lower than bad risk


## Part B
<br>
Aside questions mentioned on Part A, create 10 more questions and answers to understand more about
the dataset.

**1. Check if there is any missing values on the dataset**

In [148]:
# Reload the dataset because we already fill the nan values on Checking account column
df_gc = pd.read_csv('./german_credit_data_with_risk.csv', index_col='idx')
display(df_gc.head())

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
idx                                                                             
0     67    male    2     own             NaN           little           1169   
1     22  female    2     own          little         moderate           5951   
2     49    male    1     own          little              NaN           2096   
3     45    male    2    free          little           little           7882   
4     53    male    2    free          little           little           4870   

     Duration              Purpose  Risk  
idx                                       
0           6             radio/TV  good  
1          48             radio/TV   bad  
2          12            education  good  
3          42  furniture/equipment  good  
4          24                  car   bad

In [150]:
# Check if there is any missing values
display(df_gc.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 0 to 999
Data columns (total 10 columns):
Age                 1000 non-null int64
Sex                 1000 non-null object
Job                 1000 non-null int64
Housing             1000 non-null object
Saving accounts     817 non-null object
Checking account    606 non-null object
Credit amount       1000 non-null int64
Duration            1000 non-null int64
Purpose             1000 non-null object
Risk                1000 non-null object
dtypes: int64(4), object(6)
memory usage: 85.9+ KB


None

Looks like there are some values that missing on columns Saving accounts and Checking account

**2. How to handle the missing values from the data**
<br>
<br>
We can fill the missing values, in this case let's fill the missing values for each columns with the value `zero` because it is the case that the customer didn't have the account or simply their account balance is 0 DM

In [153]:
# Fill the nan values
df_gc['Checking account'] = df_gc['Checking account'].fillna('zero')
df_gc['Saving accounts'] = df_gc['Saving accounts'].fillna('zero')

In [156]:
# Verify the results
display(df_gc['Checking account'].value_counts())
display(df_gc['Saving accounts'].value_counts())

zero        394
little      274
moderate    269
rich         63
Name: Checking account, dtype: int64

little        603
zero          183
moderate      103
quite rich     63
rich           48
Name: Saving accounts, dtype: int64

Looks good!

**3. Check the number of unique values for each columns**

In [157]:
display(df_gc.nunique())

Age                  53
Sex                   2
Job                   4
Housing               3
Saving accounts       5
Checking account      4
Credit amount       921
Duration             33
Purpose               8
Risk                  2
dtype: int64

Looks good, but notice that the Age and Duration column have much unique values

**4. How can we handle the Age column**
<br>
<br>
We can do some binning of the ages by some ranges

In [159]:
# Bin the ages to four categories
age_cats = ['Student', 'Young', 'Adult', 'Senior']
interval = (18, 25, 35, 60, 120)
df_gc["Age_cat"] = pd.cut(df_gc.Age, interval, labels=age_cats)

In [161]:
# Check the results
display(df_gc.head())

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
idx                                                                             
0     67    male    2     own            zero           little           1169   
1     22  female    2     own          little         moderate           5951   
2     49    male    1     own          little             zero           2096   
3     45    male    2    free          little           little           7882   
4     53    male    2    free          little           little           4870   

     Duration              Purpose  Risk  Age_cat  
idx                                                
0           6             radio/TV  good   Senior  
1          48             radio/TV   bad  Student  
2          12            education  good    Adult  
3          42  furniture/equipment  good    Adult  
4          24                  car   bad    Adult

**5. How we can handle the Duration column**

In [174]:
duration_cats = ['Short', 'Middle', 'Long', 'VeryLong']
interval = (4, 12, 28, 36, 72)
df_gc["Duration_cat"] = pd.cut(df_gc.Duration, interval, labels=duration_cats)

In [175]:
# Check the results
display(df_gc.head())

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
idx                                                                             
0     67    male    2     own            zero           little           1169   
1     22  female    2     own          little         moderate           5951   
2     49    male    1     own          little             zero           2096   
3     45    male    2    free          little           little           7882   
4     53    male    2    free          little           little           4870   

     Duration              Purpose  Risk  Age_cat  installment Duration_cat  
idx                                                                          
0           6             radio/TV  good   Senior   194.833333        Short  
1          48             radio/TV   bad  Student   123.979167     VeryLong  
2          12            education  good    Adult   174.666667        Short  
3          42  furniture/equipment  good    Adult   187.666667     VeryLong  
4          24                  car   bad    Adult   202.916667       Middle

**5. How much each customer has to pay for each month**

In [176]:
# Divide the amount by the duration
df_gc['installment'] = df_gc['Credit amount']/df_gc['Duration']

In [177]:
display(df_gc.head())

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
idx                                                                             
0     67    male    2     own            zero           little           1169   
1     22  female    2     own          little         moderate           5951   
2     49    male    1     own          little             zero           2096   
3     45    male    2    free          little           little           7882   
4     53    male    2    free          little           little           4870   

     Duration              Purpose  Risk  Age_cat  installment Duration_cat  
idx                                                                          
0           6             radio/TV  good   Senior   194.833333        Short  
1          48             radio/TV   bad  Student   123.979167     VeryLong  
2          12            education  good    Adult   174.666667        Short  
3          42  furniture/equipment  good    Adult   187.666667     VeryLong  
4          24                  car   bad    Adult   202.916667       Middle

Looks good

**6. Average the credit amount based on their purpose**

In [178]:
# Grouping and sorting based on higher average amounts
display(df_gc.groupby(by='Purpose').mean()[['Credit amount']].sort_values('Credit amount', ascending=False))

Credit amount
Purpose                           
vacation/others        8209.333333
business               4158.041237
car                    3768.192878
furniture/equipment    3066.988950
education              2879.203390
repairs                2728.090909
radio/TV               2487.653571
domestic appliances    1498.000000

**7. Average the credit amount based on their duration**

In [184]:
# Grouping and sorting based on higher average amounts
display(df_gc.groupby(by='Duration_cat').mean()[['Credit amount']].sort_values('Credit amount', ascending=False))

Credit amount
Duration_cat               
VeryLong        7645.770115
Long            5326.769841
Middle          3001.390187
Short           1813.968839

In [182]:
# Display the counts of the duration
display(df_gc['Duration_cat'].value_counts())

Middle      428
Short       353
Long        126
VeryLong     87
Name: Duration_cat, dtype: int64

Based on above results we know that longer duration means more credit amounts

**8. Show top three customers that pay the most each month**

In [186]:
display(df_gc.sort_values(by='installment', ascending=False)[0:3])

Age   Sex  Job Housing Saving accounts Checking account  Credit amount  \
idx                                                                           
917   68  male    3     own          little           little          14896   
236   23  male    0     own            zero         moderate          14555   
738   45  male    3     own          little             zero           6761   

     Duration Purpose  Risk  Age_cat  installment Duration_cat  
idx                                                             
917         6     car   bad   Senior  2482.666667        Short  
236         6     car   bad  Student  2425.833333        Short  
738         6     car  good    Adult  1126.833333        Short

**9. Show the installment distribution by their risk**

In [208]:
display(df_gc.groupby(by='Risk')[['installment']].describe())

installment                                                            \
           count        mean         std        min        25%         50%   
Risk                                                                         
bad        300.0  172.044031  223.840381  24.055556  76.197917  121.058333   
good       700.0  165.819730  110.574834  25.250000  96.906250  134.916667   

                               
             75%          max  
Risk                           
bad   196.361111  2482.666667  
good  211.975000  1126.833333

**10. Show the purpose distribution for each sex**

In [213]:
display(df_gc.groupby(by=['Sex', 'Purpose']).count()[['Age']])

Age
Sex    Purpose                 
female business              19
       car                   94
       domestic appliances    6
       education             24
       furniture/equipment   74
       radio/TV              85
       repairs                5
       vacation/others        3
male   business              78
       car                  243
       domestic appliances    6
       education             35
       furniture/equipment  107
       radio/TV             195
       repairs               17
       vacation/others        9

## Part C
<br>
<br>
Define at least 3 business problems that you can find based on the dataset!

**1. Predicting new customer risk category by their data**
<br>
<br>
We can use this data to build a predictive model that can predict customer's risk category based on their data.

**2. Evaluating the business profit**
<br>
<br>
We can use this data to evaluate the business profit, and adjust the business decision accordingly.

**3. Marketing strategy**
<br>
<br>
Based on the business decision we then can know for sure what are the criteria of customer that will be targeted by marketing team.